README.md
#Usage


### GET /count/:scope/:name

get viewcount for video `:name` in scope `:scope`.

# GET /batch

batch request. Accepts a JSON body as :
```
{
  "count":{
    "scope_1":[
      "media_1",
      "media_2"
    ],
    "scope_2":[
      "media_3"
    ]
  }
}
```

Response :
```
{
  "count":{
    "scope_1":{
      "media_1":1,
      "media_2":10
    },
    "scope_2":{
      "media_3":100
    }
  }
}
```

### GET /

(unstable)
count all views on all videos


debian/changelog
carp (0.0.4) stable; urgency=medium

  * 4 hours RAM cache to prevent heavy load on elasticsearch

 -- Sebastien Dumetz <s.dumetz@holusion.com>  Thu, 13 Apr 2017 10:31:57 +0200

carp (0.0.3) stable; urgency=medium

  * use elasticsearch module internally
  * Tested reduce/expand functions

 -- Sebastien Dumetz <s.dumetz@holusion.com>  Wed, 12 Apr 2017 17:18:00 +0200

carp (0.0.2) stable; urgency=medium

  * Add a route to handle batch count requests

 -- Sebastien Dumetz <s.dumetz@holusion.com>  Tue, 11 Apr 2017 17:22:22 +0200

carp (0.0.1) stable; urgency=medium

  * Initial release. 

 -- Sebastien Dumetz <s.dumetz@holusion.com>  Fri, 07 Apr 2017 13:44:04 +0200


test/test_cache.js
'use strict';
const expect = require("chai").expect;
const cache = require("../lib/cache");


describe("cache",function(){
  it("sort cached and uncached entries",function(){
    cache.set("foo","bar",128);
    expect(cache.getCachedValues({foo:["bar","baz"]})).to.deep.equal({
      cached:{
        foo:{bar:128}
      },
      uncached:{foo:["baz"]}
    });
  });
  it("cache bust",function(){
    cache.set("foo","bar",128,Date.now()-cache.cache_time -10);
    expect(cache.getCachedValues({foo:["bar","baz"]})).to.deep.equal({
      cached:{},
      uncached:{foo:["bar","baz"]}
    });
  });
})


test/test_index.js
const expect = require("chai").expect;


describe("")


test/test_elasticQueries.js
const expect = require("chai").expect;
const queries = require("../lib/elasticQueries");

describe("elasticQueries",function(){
  describe("reduceMsearch()",function(){
    [
      [
        "simple reduction",
        {holusion:["foo"]},
        {
          items:[{scope:"holusion",name:"foo"}],
          queries:[{}, queries.count("holusion","foo")]
        }
      ],[
        "multiple scope reduction",
        {holusion:["foo"],bar:["bar"]},
        {
          items:[
            {scope:"holusion",name:"foo"},
            {scope:"bar",name:"bar"}
          ],
          queries:[
            {},
            queries.count("holusion","foo"),
            {},
            queries.count("bar","bar")
          ]
        }
      ],
    ].forEach(function(f){
      it(`${f[0]}`,function(){
        expect(queries.reduceMsearch(queries.count,f[1])).to.deep.equal(f[2]);
      })
    })
  })
  describe("expandMsearch()",function(){
    [
      [
        "simple expansion",
        {items:[{scope:"holusion",name:"foo"},{scope:"holusion",name:"bar"}]}, //we don't use the query field
        {responses:[{hits:{total:128}},{hits:{total:256}}]},
        {holusion:{foo:128,bar:256} }
      ],
    ].forEach(function(f){
      it(`${f[0]}`,function(){
        expect(queries.expandMsearch(f[1],f[2])).to.deep.equal(f[3]);
      })
    })
  })
  describe("iterate()",function(){
    [
      [
        "from requests",
        {holusion:["foo"],bar:["bar"]},
        [["holusion","foo"],["bar","bar"]]
      ],[
        "from responses",
        {holusion:{foo:128},bar:{bar:256}},
        [["holusion","foo",128],["bar","bar",256]]
      ],
    ].forEach(function(f){
      it(`${f[0]}`,function(done){
        let count = 0;
        queries.iterate(f[1],function(a,b,c){
          expect(a).to.deep.equal(f[2][count][0]);
          expect(b).to.deep.equal(f[2][count][1]);
          if(f[2][count][2]) expect(c).to.deep.equal(f[2][count][2]);
          count ++;
          if(count == f[2].length){
            done();
          }
        });
      })
    })
  })
})


lib/cache.js
'use strict';
const store = {};
const cache_time = 1000*3600*4
function get(scope,media){
  if(!store[scope] || typeof store[scope][media] === "undefined"){
    return null;
  }
  let diff = Date.now() -  store[scope][media][1];
  //diff in milliseconds -> 4 hours
  if( cache_time < diff ){
    //don't remove cache-busted items as they *will* get updated.
    return null;
  }

  return store[scope][media][0];
}

//forced_date is used for testing
function set(scope,media,value,forced_date){
  if(typeof store[scope] == "undefined"){ store[scope]={}};
  store[scope][media] = [value, forced_date || Date.now()];
}

function getCachedValues(list){
  return Object.keys(list).reduce(function(results,scope){
    list[scope].forEach(function(media){
      let views = get(scope,media);
      if(views != null){
        results.cached[scope] = results.cached[scope]||{};
        results.cached[scope][media] = views;
      }else{
        results.uncached[scope] = results.uncached[scope]||[];
        results.uncached[scope].push(media);
      }
    });
    return results;
  },{cached:{},uncached:{}});
}

module.exports = {get, set, getCachedValues, cache_time};


lib/elasticQueries.js
'use strict';

function count(scope,media){
  return {
    query: {
      query_string: {
        query: `request: "/playlist/${encodeURIComponent(scope)}/${encodeURIComponent(media)}.mp4" AND (response: 200 OR (response:206  AND offset: 0) OR response: 304)`,
        "analyze_wildcard": true
      }
    }
  }
}

//Flatten a batch query into an array of sources and an array of queries.
//see tests for formats
function reduceMsearch(op, items){
  return Object.keys(items).reduce(function(r,scope){
    r.items = r.items.concat(items[scope].map(function(media){
      return {name:media,scope:scope};
    }))
    r.queries = r.queries.concat(items[scope].reduce(function(ar, media){
      ar.push({});
      ar.push(op(scope,media));
      return ar;
    },[]));
    return r;
  },{items:[],queries:[]});
}

function expandMsearch(q, r){
  const responses = r.responses;
  if(q.items.length != responses.length){
    throw new Error("queries and responses length must match");
  }

  return q.items.reduce(function(out,item,index){
    //create scope if it doesn't exist
    let scope_res = out[item.scope]||{};
    if(responses[index] && responses[index].hits){
      scope_res[item.name] = responses[index].hits.total;
    }else{
      scope_res[item.name] = 0;
      console.warn("Invalid response : ",responses[index]);
    }

    out[item.scope] = scope_res;
    return out;
  },{});
}
// sequence of fn(scope,media,views) or fn(scope,media) depending if it's called with requests or results.
function iterate(batch,fn){
  Object.keys(batch).forEach(function(scope){
    if(Array.isArray(batch[scope])){
      batch[scope].forEach(function(media){
        fn(scope,media);
      })
    }else{
      Object.keys(batch[scope]).forEach(function(media){
        fn(scope,media,batch[scope][media]);
      })
    }
  })
}
module.exports = {reduceMsearch, expandMsearch, count, iterate};


lib/index.js
'use strict';
const http = require('http');
const jsonBodyParser = require("body-parser").json();
const express = require("express");
const elasticsearch = require('elasticsearch');
const queries = require("./elasticQueries");
const cache = require("./cache");

function countAll(callback){
  const q = {
    query: {
      query_string: {
        query: `request: /playlist/*.mp4 AND (response: 200 OR (response:206  AND offset: 0) OR response: 304)`,
        "analyze_wildcard": true
      }
    }
  }
  client.search(q,function(err,res){
    if(err != null){
      return callback(err);
    }
    return callback(null,res.hits.total);
  });
}

function batchCount(things,callback){

  //sort cached / uncached requests
  const requests = cache.getCachedValues(things);
  //Serialize. We use an intermediate object to find back scope & name after elasticsearch answers
  const batch = queries.reduceMsearch(queries.count, requests.uncached);
  //Extract queries for mquery()

  function parseResults(err,data){
    if(err){
      return callback(err);
    }
    //Use this to debug future ElasticSearch response syntax changes
    //console.log("ES Response Raw : ", data);
    const responses = data.responses;
    //console.log("Raw results :",data.responses);
    //Deserialize to find results structure
    const formatted_results = queries.expandMsearch(batch,data);
    //Merge cached results and new queries. Add new queries to cache.
    queries.iterate(formatted_results,function(scope,media,views){
      cache.set(scope,media,views);
      if(!requests.cached[scope]){
        requests.cached[scope] = {};
      }
      requests.cached[scope][media]=views;
    });
    callback(null,requests.cached);
  };
  if(0 < batch.queries.length){
    client.msearch({body:batch.queries},parseResults)
  }else{
    parseResults(null,{responses:[]});
  }
}


function count(scope, media, callback){
  const q = queries.count(scope,media);
  client.search(q,function(err,res){
    if(err != null){
      return callback(err);
    }
    return callback(null,res.hits.total);
  });
}


function enableCORS(req, res, next) {
  res.header("Access-Control-Allow-Origin", "*"/*"https://pixel.holusion.com"*/);
  res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
  next();
};

// BOOTSTRAP //
const client = new elasticsearch.Client({
  apiVersion: "5.0",
  host: 'localhost:9200',
  log: 'info'
});

client.ping({
  requestTimeout: 10000,
}, function (error) {
  if (error) {
    console.error('elasticsearch cluster is down!');
  } else {
    console.log('All is well');
  }
});

const app = express();
app.use(enableCORS);
app.get("/",function(req, res){
  countAll(function(err,count){
    if( err != null ){
      return res.status(err.status||500).send(
        (err.root_cause)?JSON.stringify(err.root_cause[0]):err
      );
    }
    res.set("Content-Type","text/plain");
    res.send(count.toString());
  });
});
app.post("/batch",jsonBodyParser, function(req,res){
  if(typeof req.body.count != "object"){
    return res.status(400).send({code:400,message:"Request must have a count field"});
  }
  batchCount(req.body.count,function(err, data){
    if(err){
      return res.status(err.status||500).send(
        (err.root_cause)?JSON.stringify(err.root_cause[0]):err
      );
    }
    res.set("Content-Type","application/json");
    res.send({count:data});
  });
});
app.get("/count/:scope/:media",function(req,res){
  count(req.params.scope,req.params.media,function(err,count){
    if(err){
      return res.status(err.status||500).send(
        (err.root_cause)?JSON.stringify(err.root_cause[0]):err
      );
    }
    res.set("Content-Type","text/plain");
    res.send(count.toString());
  });
});
const port = process.env["PORT"] || 3000;
app.listen(port, function () {
  console.log(`app listening on port ${port}!`);
})
